torchgeo/trainers/byol.py

```diff
@@ -0,0 +1,393 @@
+
+import torch
+import torch.nn.functional as F
+from kornia import augmentation as K
+from kornia import filters
+from kornia.geometry import transform as KorniaTransform
+from pytorch_lightning.core.lightning import LightningModule
+from torch import Tensor, optim
+from torch.autograd import Variable
+from torch.nn.modules import BatchNorm1d, Conv2d, Linear, Module, ReLU, Sequential
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+from torch.utils.data import DataLoader
+from torchvision.models import resnet18
+from torchvision.models.resnet import resnet50
+
+DataLoader.__module__ = "torch.utils.data"
+Module.__module__ = "torch.nn"
+Sequential.__module__ = "torch.nn"
+Linear.__module__ = "torch.nn"
+ReLU.__module__ = "torch.nn"
+BatchNorm1d.__module__ = "torch.nn"
+Conv2d.__module__ = "torch.nn"
+
+
+def simCLR_default_augmentation(image_size: Tuple[int, int] = (256, 256)) -> Module:
+    """Applies default augmentation from simCLR.
+
+    Args:
+        image_size: Tuple of integers defining the image size
+    """
+    return Sequential(
+        KorniaTransform.Resize(size=image_size),  # type: ignore[attr-defined]
+        # Not suitable for multispectral adapt
+        # RandomApply(K.ColorJitter(0.8, 0.8, 0.8, 0.2), p=0.8),
+        # K.RandomGrayscale(p=0.2),
+        K.RandomHorizontalFlip(),
+        RandomApply(filters.GaussianBlur2d((3, 3), (1.5, 1.5)), p=0.1),
+        K.RandomResizedCrop(size=image_size),
+    )
+
+
+def normalized_mse(x: Tensor, y: Tensor) -> Tensor:
+    """Computes the normalized mean square error between x and y.
+
+    Args:
+        x: Tensor x
+        y: Tensor y
+    """
+    x = F.normalize(x, dim=-1)
+    y = F.normalize(y, dim=-1)
+    mse = torch.mean(2 - 2 * (x * y).sum(dim=-1))  # type: ignore[attr-defined]
+    return cast(Tensor, mse)
+
+
+def mlp(dim: int, projection_size: int = 256, hidden_size: int = 4096) -> Module:
+    """MLP used in the projection head.
+
+    Args:
+        dim: size of layer to project
+        projection_size: First layer MLP projection size
+        hidden_size: Size of MLP hidden layer
+    """
+    return Sequential(
+        Linear(dim, hidden_size),
+        BatchNorm1d(hidden_size),  # type: ignore[no-untyped-call]
+        ReLU(inplace=True),
+        Linear(hidden_size, projection_size),
+    )
+
+
+class RandomApply(Module):
+    """Applies augmentation function (augm) with probability p."""
+
+    def __init__(
+        self, augm: Callable[[Dict[str, Tensor]], Dict[str, Tensor]], p: float
+    ) -> None:
+        """Initialize RandomApply.
+
+        Keyword Args:
+            augm: Augmentation finction to aply
+            p: Probability with wich the augmentation (augm) fn is applied
+        """
+        super().__init__()  # type: ignore[no-untyped-call]
+        self.augm = augm
+        self.p = p
+
+    def forward(self, x: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        """Randomapply forward method."""
```

Comment:
This docstring isn't very descriptive and doesn't mention input args or return values